File: Graph/python/longestPath.py
from .graph_utils import Graph


def longest_path(graph: Graph) -> int:
    distance: dict[str, int] = {node: 0 for node, neighbors in graph.items() if not neighbors}
    best = 0

    for node in graph:
        value = _traverse(graph, node, distance)
        best = max(best, value)

    return best


def _traverse(graph: Graph, node: str, distance: dict[str, int]) -> int:
    if node in distance:
        return distance[node]

    max_neighbor = 0
    for neighbor in graph.get(node, []):
        attempt = _traverse(graph, neighbor, distance)
        max_neighbor = max(max_neighbor, attempt)

    distance[node] = max_neighbor + 1
    return distance[node]

File: Graph/java/LongestPath.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LongestPath {
  public static int longest(Map<String, List<String>> graph) {
    Map<String, Integer> distance = new HashMap<>();
    for (Map.Entry<String, List<String>> entry : graph.entrySet()) {
      if (entry.getValue().isEmpty()) {
        distance.put(entry.getKey(), 0);
      }
    }

    int best = 0;
    for (String node : graph.keySet()) {
      int value = traverse(graph, node, distance);
      if (value > best) {
        best = value;
      }
    }
    return best;
  }

  private static int traverse(Map<String, List<String>> graph, String node, Map<String, Integer> distance) {
    if (distance.containsKey(node)) {
      return distance.get(node);
    }

    int maxNeighbor = 0;
    for (String neighbor : graph.getOrDefault(node, List.of())) {
      int candidate = traverse(graph, neighbor, distance);
      if (candidate > maxNeighbor) {
        maxNeighbor = candidate;
      }
    }

    distance.put(node, maxNeighbor + 1);
    return distance.get(node);
  }
}
